pkg/block/block.go

```diff
@@ -89,41 +92,66 @@ func Upload(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bdir st
 	}
 
 	if err := objstore.UploadDir(ctx, logger, bkt, path.Join(bdir, ChunksDirname), path.Join(id.String(), ChunksDirname)); err != nil {
-		return cleanUp(bkt, id, errors.Wrap(err, "upload chunks"))
+		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload chunks"))
 	}
 
 	if err := objstore.UploadFile(ctx, logger, bkt, path.Join(bdir, IndexFilename), path.Join(id.String(), IndexFilename)); err != nil {
-		return cleanUp(bkt, id, errors.Wrap(err, "upload index"))
+		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload index"))
 	}
 
 	if meta.Thanos.Source == metadata.CompactorSource {
 		if err := objstore.UploadFile(ctx, logger, bkt, path.Join(bdir, IndexCacheFilename), path.Join(id.String(), IndexCacheFilename)); err != nil {
-			return cleanUp(bkt, id, errors.Wrap(err, "upload index cache"))
+			return cleanUp(logger, bkt, id, errors.Wrap(err, "upload index cache"))
 		}
 	}
 
 	// Meta.json always need to be uploaded as a last item. This will allow to assume block directories without meta file
 	// to be pending uploads.
 	if err := objstore.UploadFile(ctx, logger, bkt, path.Join(bdir, MetaFilename), path.Join(id.String(), MetaFilename)); err != nil {
-		return cleanUp(bkt, id, errors.Wrap(err, "upload meta file"))
+		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload meta file"))
 	}
 
 	return nil
 }
 
-func cleanUp(bkt objstore.Bucket, id ulid.ULID, err error) error {
+func cleanUp(logger log.Logger, bkt objstore.Bucket, id ulid.ULID, err error) error {
 	// Cleanup the dir with an uncancelable context.
-	cleanErr := Delete(context.Background(), bkt, id)
+	cleanErr := Delete(context.Background(), logger, bkt, id)
 	if cleanErr != nil {
 		return errors.Wrapf(err, "failed to clean block after upload issue. Partial block in system. Err: %s", err.Error())
 	}
 	return err
 }
 
-// Delete removes directory that is mean to be block directory.
-// NOTE: Prefer this method instead of objstore.Delete to avoid deleting empty dir (whole bucket) by mistake.
-func Delete(ctx context.Context, bucket objstore.Bucket, id ulid.ULID) error {
-	return objstore.DeleteDir(ctx, bucket, id.String())
+// Delete removes directory that is meant to be block directory.
+// NOTE: Always prefer this method for deleting blocks.
+//  * We have to delete block's files in the certain order (meta.json first)
+//  to ensure we don't end up with malformed partial blocks. Thanos system handles well partial blocks
+//  only if they don't have meta.json. If meta.json is present Thanos assumes valid block.
+//  * This avoids deleting empty dir (whole bucket) by mistake.
+func Delete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, id ulid.ULID) error {
+	if err := bkt.Delete(ctx, path.Join(id.String(), MetaFilename)); err != nil {
```

Comment:
Maybe it makes sense to put `path.Join(id.String(), MetaFilename)` into a variable, to avoid mismatches in the future with the second time below that you use this path.

Reply:
:+1: 
Fixed